ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplExternalTables.java

```diff
@@ -221,7 +251,66 @@ private void dirLocationToCopy(FileList fileList, Path sourcePath, HiveConf conf
         targetPath = new Path(Utils.replaceHost(targetPath.toString(), sourcePath.toUri().getHost()));
         sourcePath = new Path(Utils.replaceHost(sourcePath.toString(), remoteNS));
       }
-      fileList.add(new DirCopyWork(sourcePath, targetPath).convertToString());
+      fileList.add(
+          new DirCopyWork(sourcePath, targetPath, copyMode).convertToString());
+    }
+
+    private SnapshotUtils.SnapshotCopyMode createSnapshotsAtSource(
+        Path sourcePath, String snapshotPrefix, boolean isSnapshotEnabled,
+        HiveConf conf, HashSet<String> numFailedSnapshotPaths) {
+      if (!isSnapshotEnabled) {
+        LOG.info("Snapshot copy not enabled for path {} Will use normal "
+                + "distCp for copying data.",
+            sourcePath);
+        return FALLBACK_COPY;
+      }
+      DistributedFileSystem sourceDfs = SnapshotUtils.getDFS(sourcePath, conf);
+
+      if (sourceDfs != null) {
+        try {
+          // check if second snapshot exists.
+          boolean isSecondSnapAvlb = SnapshotUtils
+              .isSnapshotAvailable(sourceDfs, sourcePath, snapshotPrefix,
+                  "old");
+          if (isSecondSnapAvlb) {
+            sourceDfs.deleteSnapshot(sourcePath, snapshotPrefix + "old");
+            sourceDfs.renameSnapshot(sourcePath, snapshotPrefix + "initial",
+                snapshotPrefix + "old");
+            sourceDfs.createSnapshot(sourcePath, snapshotPrefix + "initial");
+            return DIFF_COPY;
+          } else {
+            // Check if first snapshot is available
+            boolean isFirstSnapshotAvailable = SnapshotUtils
+                .isSnapshotAvailable(sourceDfs, sourcePath, snapshotPrefix,
+                    "initial");
+            if (isFirstSnapshotAvailable) {
+              sourceDfs.renameSnapshot(sourcePath, snapshotPrefix + "initial",
+                  snapshotPrefix + "old");
+              sourceDfs.createSnapshot(sourcePath, snapshotPrefix + "initial");
+              return DIFF_COPY;
+            } else {
+              if (SnapshotUtils.allowSnapshot(sourceDfs, sourcePath)) {
+                ReplUtils.createSnapshotSafe(sourceDfs, sourcePath,
+                    snapshotPrefix + "initial");
+                return INITIAL_COPY;
+              } else {
+                numFailedSnapshotPaths.add(sourcePath.toString());
+                LOG.error("Can not allow Snapshot for path {}", sourcePath);
+                return FALLBACK_COPY;
+              }
+            }
+          }
+        } catch (Exception e) {
+          LOG.error("Error encountered during snapshot setup for path {}",
+              sourcePath, e);
+          numFailedSnapshotPaths.add(sourcePath.toString());
+          return FALLBACK_COPY;
+        }
+      } else {
```

Comment:
this else is not needed. all other paths return from the top.

Reply:
This `else` is when the Source isn't dfs. for this `if`
```
    if (sourceDfs != null) {
```